### ISSUE TITLE
[FR] Support BSHM-layout scaled_dot_product_attention without transpose.

### ISSUE THREAD

**Issue by ghostplant** [pytorch/pytorch#152257]
What's the plan to support direct computation given (Batch, Seq, Head, Model_dim) Q/K/V tensors, without additional expensive back-and-forth?

```sh
q = torch.randn([b, s, h, m])
k = torch.randn([b, s, h, m])
v = torch.randn([b, s, h, m])

scores = torch.nn.functional.scaled_dot_product_attention(q, k, v, no_transpose=True)
```

**Comment by AbhiLegend** [comment]
I have used an alternative way to deal with the issue(I have tested in Colab GPU) as the features still to come this might be a way.
`import torch

def scaled_dot_product_attention_bshd(q, k, v, is_causal=False, dropout_p=0.0):
    """
    q, k, v: (batch, seq, head, dim)
    """

    # Reinterpret Q/K/V as (batch, head, seq, dim) without copying
    q_ = q.permute(0, 2, 1, 3)  # (B, H, S, D)
    k_ = k.permute(0, 2, 1, 3)  # (B, H, S, D)
    v_ = v.permute(0, 2, 1, 3)  # (B, H, S, D)

    # Now safe to call
    output = torch.nn.functional.scaled_dot_product_attention(
        q_,
        k_,
        v_,
        is_causal=is_causal,
        dropout_p=dropout_p,
    )

    # output is (B, H, S, D) — convert back if needed
    output = output.permute(0, 2, 1, 3)  # (B, S, H, D)

    return output

 b, s, h, d = 2, 8, 4, 16
q = torch.randn(b, s, h, d)
k = torch.randn(b, s, h, d)
v = torch.randn(b, s, h, d)

out = scaled_dot_product_attention_bshd(q, k, v)
print(out.shape)  # (B, S, H, D)

`
I am attaching the Google COLAB notebook with benchmark results [https://colab.research.google.com/drive/159bHVgXVsQwTnN1up3MoC0HIJ27WhxbA?usp=sharing](url)

**Comment by AbhiLegend** [comment]
I think there might be some future development coming up this way.It's an assumption.It might look something like this
`scores = scaled_dot_product_attention(q, k, v, layout="bshd")`.
These following things might be done in future

- Plan is to add a layout argument (like layout="bshd")
- Kernels will natively compute attention with (Batch, Seq, Head, Dim) format
- No extra transpose needed in future
- Faster, cleaner, less memory overhead

**Comment by AbhiLegend** [comment]
This is my first contribution to Pytorch Repo if the things are good  can I work on it  @ghostplant  .Please review the notebook too.

**Comment by ghostplant** [comment]
> I have used an alternative way to deal with the issue(I have tested in Colab GPU) as the features still to come this might be a way. `import torch
> 
> def scaled_dot_product_attention_bshd(q, k, v, is_causal=False, dropout_p=0.0): """ q, k, v: (batch, seq, head, dim) """
> 
> ```
> # Reinterpret Q/K/V as (batch, head, seq, dim) without copying
> q_ = q.permute(0, 2, 1, 3)  # (B, H, S, D)
> k_ = k.permute(0, 2, 1, 3)  # (B, H, S, D)
> v_ = v.permute(0, 2, 1, 3)  # (B, H, S, D)
> 
> # Now safe to call
> output = torch.nn.functional.scaled_dot_product_attention(
>     q_,
>     k_,
>     v_,
>     is_causal=is_causal,
>     dropout_p=dropout_p,
> )
> 
> # output is (B, H, S, D) — convert back if needed
> output = output.permute(0, 2, 1, 3)  # (B, S, H, D)
> 
> return output
> ```
> 
> b, s, h, d = 2, 8, 4, 16 q = torch.randn(b, s, h, d) k = torch.randn(b, s, h, d) v = torch.randn(b, s, h, d)
> 
> out = scaled_dot_product_attention_bshd(q, k, v) print(out.shape) # (B, S, H, D)
> 
> ` I am attaching the Google COLAB notebook with benchmark results https://colab.research.google.com/drive/159bHVgXVsQwTnN1up3MoC0HIJ27WhxbA?usp=sharing

Is it really efficient (I mean no extra expensive element-wise transpose operations)?

**Comment by AbhiLegend** [comment]
> > I have used an alternative way to deal with the issue(I have tested in Colab GPU) as the features still to come this might be a way. `import torch
> > def scaled_dot_product_attention_bshd(q, k, v, is_causal=False, dropout_p=0.0): """ q, k, v: (batch, seq, head, dim) """
> > ```
> > # Reinterpret Q/K/V as (batch, head, seq, dim) without copying
> > q_ = q.permute(0, 2, 1, 3)  # (B, H, S, D)
> > k_ = k.permute(0, 2, 1, 3)  # (B, H, S, D)
> > v_ = v.permute(0, 2, 1, 3)  # (B, H, S, D)
> > 
> > # Now safe to call
> > output = torch.nn.functional.scaled_dot_product_attention(
> >     q_,
> >     k_,
> >     v_,
> >     is_causal=is_causal,
> >     dropout_p=dropout_p,
> > )
> > 
> > # output is (B, H, S, D) — convert back if needed
> > output = output.permute(0, 2, 1, 3)  # (B, S, H, D)
> > 
> > return output
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > b, s, h, d = 2, 8, 4, 16 q = torch.randn(b, s, h, d) k = torch.randn(b, s, h, d) v = torch.randn(b, s, h, d)
> > out = scaled_dot_product_attention_bshd(q, k, v) print(out.shape) # (B, S, H, D)
> > ` I am attaching the Google COLAB notebook with benchmark results https://colab.research.google.com/drive/159bHVgXVsQwTnN1up3MoC0HIJ27WhxbA?usp=sharing
> 
> Is it really efficient (I mean no extra expensive element-wise transpose operations)?

o avoid unnecessary memory copy operations while adapting (B, S, H, D) input format for scaled_dot_product_attention, we use efficient .permute() operations which reinterpret strides without performing data movement, confirmed by benchmarks.

**Comment by AbhiLegend** [comment]
> > > I have used an alternative way to deal with the issue(I have tested in Colab GPU) as the features still to come this might be a way. `import torch
> > > def scaled_dot_product_attention_bshd(q, k, v, is_causal=False, dropout_p=0.0): """ q, k, v: (batch, seq, head, dim) """
> > > ```
> > > # Reinterpret Q/K/V as (batch, head, seq, dim) without copying
> > > q_ = q.permute(0, 2, 1, 3)  # (B, H, S, D)
> > > k_ = k.permute(0, 2, 1, 3)  # (B, H, S, D)
> > > v_ = v.permute(0, 2, 1, 3)  # (B, H, S, D)
> > > 
> > > # Now safe to call
> > > output = torch.nn.functional.scaled_dot_product_attention(
> > >     q_,
> > >     k_,
> > >     v_,
> > >     is_causal=is_causal,
> > >     dropout_p=dropout_p,
> > > )
> > > 
> > > # output is (B, H, S, D) — convert back if needed
> > > output = output.permute(0, 2, 1, 3)  # (B, S, H, D)
> > > 
> > > return output
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > b, s, h, d = 2, 8, 4, 16 q = torch.randn(b, s, h, d) k = torch.randn(b, s, h, d) v = torch.randn(b, s, h, d)
> > > out = scaled_dot_product_attention_bshd(q, k, v) print(out.shape) # (B, S, H, D)
> > > ` I am attaching the Google COLAB notebook with benchmark results https://colab.research.google.com/drive/159bHVgXVsQwTnN1up3MoC0HIJ27WhxbA?usp=sharing
> > 
> > 
> > Is it really efficient (I mean no extra expensive element-wise transpose operations)?
> 
> To avoid unnecessary memory copy operations while adapting (B, S, H, D) input format for scaled_dot_product_attention, we use efficient .permute() operations which reinterpret strides without performing data movement, confirmed by benchmarks.

@ghostplant  If you have time, you can check and if possible assign the issue to me :-)

**Comment by ghostplant** [comment]
@AbhiLegend Yes, it seems work, but actually I just found another issue, this builtin interface doesn't have num_kvsplit, so it executes significantly slow when S turns large (e.g. 4096).

**Comment by ghostplant** [comment]
Should be also an important extended argument to support and configure.

**Comment by AbhiLegend** [comment]
> [@AbhiLegend](https://github.com/AbhiLegend) Yes, it seems work, but actually I just found another issue, this builtin interface doesn't have num_kvsplit, so it executes significantly slow when S turns large (e.g. 4096).

I have update the following changes in the notebook please see @ghostplant  if it works for you

**Major Enhancements Made to Scaled Dot-Product Attention**:

> Added support for configurable and automatic key/value splitting via num_kv_split
> 
> Implemented auto-tuning based on sequence length and device
> 
> Preserved mathematical correctness (0 max diff from baseline)
> 
> Introduced full benchmark suite with plotting
> 
> Prepared for scaling attention to sequence lengths ≥ 8192 efficiently

https://colab.research.google.com/drive/159bHVgXVsQwTnN1up3MoC0HIJ27WhxbA?usp=sharing

**Comment by AbhiLegend** [comment]
> Should be also an important extended argument to support and configure.

Can you please assign this issue to me @ghostplant

**Comment by AbhiLegend** [comment]
> [@AbhiLegend](https://github.com/AbhiLegend) Of course, thank you for your plans! We expect a "kvsplit" support in torch builtin SDPA, instead of manually chunking all the inputs and doing the concat, which may not be correct in results.

The issue is still unassigned please sign it to me @ghostplant

**Comment by AbhiLegend** [comment]
> Sorry, I don't have the permission in changing the assignment.

No problem let me find a way.
